v3-roadmap-and-design.md

```diff
@@ -0,0 +1,425 @@
+
+### Creation API
+
+Zarr-Python 2.* bundles together the creation and serialization of Zarr objects. Zarr-Python 3.* will make it possible to create objects in memory separate from serializing them. This will specifically enable writing hierarchies of Zarr objects in a single batch step. For example:
+
+```python
+
+arr1 = Array(shape=(10, 10), path="foo/bar", dtype="i4", store=store)
+arr2 = Array(shape=(10, 10), path="foo/spam", dtype="f8", store=store)
+
+arr1.save()
+arr2.save()
+
+# or equivalently
+
+zarr.save_many([arr1 ,arr2])
+```
+
+### Plugin API
+
+Zarr V3 was designed to be extensible at multiple layers. Zarr-Python will support these extensions through a combination of [Abstract Base Classes](https://docs.python.org/3/library/abc.html) (ABCs) and [Entrypoints](https://packaging.python.org/en/latest/specifications/entry-points/). 
+
+**ABCs**
+
+Zarr V3 will expose Abstract base classes for the following objects:
+
+- `Store`, `ReadStore`, `ReadWriteStore`, `ReadListStore`, and `ReadWriteListStore`
+- `BaseArray`, `SynchronousArray`, and `AsynchronousArray`
+- `BaseGroup`, `SynchronousGroup`, and `AsynchronousGroup`
+- `Codec`, `ArrayArrayCodec`, `ArrayBytesCodec`, `BytesBytesCodec`
+
+**Entrypoints**
+
+Lots more thinking here but the idea here is to provide entrypoints for `data type`, `chunk grid`, `chunk key encoding`, `codecs`,  `storage_transformers` and `stores`.  These might look something like:
+
+```
+entry_points="""
+    [zarr.codecs]
+    blosc_codec=codec_plugin:make_blosc_codec
+    zlib_codec=codec_plugin:make_zlib_codec
+"""
+```
+
+### Python type hints and static analysis
+
+Target 100% Mypy coverage in 3.0 source. 
+
+### Observability
+
+A persistent problem in Zarr-Python is diagnosing problems that span many parts of the stack. To address this in 3.0, we will add a basic logging framework that can be used to debug behavior at various levels of the stack. We propose to add the separate loggers for the following namespaces:
+
+- `array`
+- `group`
+- `store`
+- `codec`
+
+These should be documented such that users know how to activate them and developers know how to use them when developing extensions.
+
+### Dependencies
+
+Today, Zarr-Python has the following required dependencies:
+
+```python
+dependencies = [
+    'asciitree',
+    'numpy>=1.20,!=1.21.0',
+    'fasteners',
+    'numcodecs>=0.10.0',
+]
+```
+
+What other dependencies should be considered?
+
+1. Attrs - Zarrita makes extensive use of the Attrs library
+2. Fsspec - Zarrita has a hard dependency on Fsspec. This could be easily relaxed though.
+
+## Breaking changes relative to Zarr-Python 2.*
+
+1. H5py compat moved to a stand alone module?
+2. `Group.__getitem__` support moved to `Group.members.__getitem__`?
+3. Others? 
```

Comment:
This is really important point @jeromekelleher! The doc says there will be breaking changes to APIs, but doesn't specify what those changes will be. In order for the community to get on board, we should be super clear about this. My hope is that breaking changes are very minimal, and whatever breaking changes there are should be justified with a clearly argued rationale.

Reply:
To be honest, we're still sorting out the feasibility of maintaining API compatibility at the Array level. @d-v-b has been leading this so I'll let him share his thoughts.

Reply:
There are a couple fundamental challenges for preserving the `zarr-python` v2 api after we fully integrate v3: The `Array` class in `zarr-python` exposes all the attributes defined for zarr version 2 arrays. But zarr version 3 arrays have different attributes, so in `zarr-python` version 3 we have a few options for dealing with this: 
1. Uses the current `zarr-python` `Array` class to wrap v3 arrays and make them look like v2 arrays. this maximizes backwards compatibility, but would be a usability nightmare and confusing for people who want to work primarily with v3 arrays. 
2. Create a new `Array` class in `zarr-python` that exposes zarr version 3 array attributes, much like `zarr-python` v2 did for zarr version 2 arrays, and then use this class to wrap zarr version 2 arrays. This would also be unpleasant, because zarr version 3 arrays don't have a `dtype` property (instead, that attribute is called `data_type`), and so hewing too close the zarr version 3 array spec for `zarr-python` would result in an impedance mismatch with the wider pydata ecosystem.
3. elevate the `Array` class to a level of abstraction above zarr version 2 and version 3 array specifications, and make that `Array` class capable of using zarr version 2 or 3 arrays as load-bearing members. This lets us optimize the `Array` class in zarr-python for a specific goal: a performant, familiar API for people coming from other pydata tools, while smoothing over the differences between v2 and v3 arrays. I think this is the best option, but it does involve substantial breaking changes to the `zarr-python` array API.

The only option here that doesn't break the `zarr-python` array API is 1, but it's a terrible option for people who want to get the most out of zarr version 3. The fact is, zarr v3 is a breaking change w.r.t. zarr v2, and I would suggest that people who do not want their code to break should continue using zarr v2 and pin `zarr-python` to version 2.

And this doesn't touch performance at all. I think we will need to change a lot of APIs to get good performance in zarr-python version 3.